pkg/driver/device_manager.go

```diff
@@ -94,11 +92,11 @@ type DeviceManager interface {
 	// GetUUID get the UuidResponse for a given uuid
 	GetUUID(uuid.UUID) ([]byte, error)
 	// GetLogsReader get the logs for a given uuid
-	GetLogsReader(u uuid.UUID) (io.Reader, error)
+	GetLogsReader(u uuid.UUID) (common.ChunkReader, error)
 	// GetInfoReader get the info for a given uuid
-	GetInfoReader(u uuid.UUID) (io.Reader, error)
+	GetInfoReader(u uuid.UUID) (common.ChunkReader, error)
 	// GetRequestsReader get the request logs for a given uuid
-	GetRequestsReader(u uuid.UUID) (io.Reader, error)
+	GetRequestsReader(u uuid.UUID) (common.ChunkReader, error)
```

Comment:
See my comment in the PR comments about the `common.ChunkReader`.

Reply:
We return reader for stored data in dir/memory/stream and propagate the data via rest api. But with change of the storing format to proto we still want to be able to present the data in proto OR json. Conversion functions convert the data to json, but it is not possible for our readers as they merge several files/slices/messages ino the stream. In case of inner proto format we cannot convert data from the stream, only from some kind of chunks (file/slice/message).

Reply:
> where the apiHandler/adminHandler reads the stream from the io.Reader and parses/chunks it as necessary?

In case of no conversion needed (before my PR we store and send the same JSON), it is ok to just do copy of such stream in adminHandler to the HTTP response writer. But in case of storing of proto, we want to be able to convert the data back to JSON in adminHandler (conditionally, now I look onto Accept header to decide, should we do conversion or not).

> the backend implementations are clueless about the data meaning and the borders between them

The backend stores the data with borders, as we have functions like `WriteInfo` or `WriteFlowMessage`, and each call writes the chunk (one message), not the bunch of them. We store every info/log/metric... in separate files/slices/messages and I changed the logic to return reader and size for the next chunk of data to be able to convert the data in adminHandler into format we requested.

Reply:
Let me see if I understand this. You are saying that the read/write are mismatched (example for info, same for logs and requests).

* write is `WriteInfo(uuid.UUID, []byte)` - this writes a single message
* read is `GetInfoReader(uuid.UUID)` - this gets a stream reader for lots of sequential messages

Because of the above, the backend implementation (too long, I will write "impl" from now on) knows the borders between messages. It might store them in a single file or []byte in a stream, it might have separate files or `[]byte` slices, we don't know, but the backend knows.

When we read, we get a stream, without the backend impl passing us any information on where the borders are.

We could fix that by doing `ReadInfo(uuid.UUID, someID)`, but that would require a _lot_ of calls, and keeping track of messages on the client, etc. So instead, we need something like: `ReadInfo()` that returns a "reader that includes border between messages". This `ChunkReader` is intended to give that "reader that includes border between messages". Is that it?

This `ChunkReader` is basically just like `tar.Reader`?

Reply:
Your understanding is correct, ChunkReader(we can discuss better naming) allow us to request the reader for the next chunk(info/log..) and from the caller we will see the borders.

Reply:
OK, now it makes sense. We definitely should include some comments to explain how this works, probably on the interface described here, so the next person knows why it works this way.

Should we use `ChunkReader`? Or should we just implement it as a `tar.Reader`? That is something everyone knows how to read?

`tar.Reader` is in my mind a lot lately from another project, so I might have a bias in that direction.

Reply:
I put comment on top of the interface.
I do not think that I can use something from tar.Reader here, as tar.Reader not an interface, but implementation. But probably I just cannot understand your suggestion.

Reply:
> changing from NextChunkReader() to Next().

Done

> move the ChunkReader interface to device_manager.go

Unfortunately it come with cyclic import, but if you really want, I can move `MaxSizes` struct and resolve the issue. Will it make sense?

> let's add the documentation in the interface in device_manager.go describing why we this is structured this way, so the next person gets it.

I modified the comment again

> when I get to the final chunk, and I call Next(), does it block? error? io.EOF?

Also mentioned in comment

Reply:
Not worth the effort. Leave it as is.